src/x86/gpio.c

```diff
@@ -0,0 +1,105 @@
+#include "gpio.h"
+
+#include <stdbool.h>
+#include <stdint.h>
+
+// TODO(ckitagawa): This whole implementation works and is fast but is bad due to the need for
+// sourcing configurations. Figure out an alternative. Compared to registers it is memory
+// inefficient but it saves having to define fake registers and stores the same information.
+
+// TODO(ckitagawa): Move this to a configuration file which is included in the x86 shared header.
+#define MAX_PORTS 6
+#define MAX_PINS 16
+
+// TODO(ckitagawa): Move to a shared header to make this publically available to share between all
+// x86 HAL implementations. They can then read from these to get information on current settings.
+static GPIOSettings pin_settings[MAX_PORTS * MAX_PINS];
+uint8_t gpio_x86_pin_input_value[MAX_PORTS * MAX_PINS];
+
+// Determines if an GPIOAddress is valid based on the defined number of ports and pins.
+static inline bool prv_is_address_valid(const GPIOAddress *address) {
+  if (address->port >= MAX_PORTS || address->pin >= MAX_PINS) {
+    return false;
+  }
+  return true;
+}
+
+// Determines if a GPIOState is valid based on the enums.
+static inline bool prv_is_state_valid(const GPIOState *state) {
+  if (*state >= NUM_GPIO_STATE) {
+    return false;
+  }
+  return true;
+}
+
+// Determines if a GPIOSettings is valid based on the enums.
+static inline bool prv_are_settings_valid(const GPIOSettings *settings) {
+  if (settings->direction >= NUM_GPIO_DIR || settings->state >= NUM_GPIO_STATE ||
+      settings->resistor >= NUM_GPIO_RES || settings->alt_function >= NUM_GPIO_ALTFN) {
+    return false;
+  }
+  return true;
+}
+
+static inline uint32_t prv_get_index(GPIOAddress *address) {
+  return address->port * MAX_PORTS + address->pin;
```

Comment:
Again, you should leave the inlining to the compiler. Anything else screams premature optimization at this point.